Simplenote/NoticeController.swift

```diff
@@ -57,11 +51,47 @@ class NoticeController {
             return
         }
 
-        if notice == current {
+        notices.append(notice)
+    }
+
+    private func dismissNoticeIfNeeded(completion: (() -> Void)?) {
+        switch state {
+        case .inactive:
+            presentNextIfPossible()
+        case .presenting:
+            dismiss(withDuration: UIKitConstants.animationQuickDuration, completion: completion)
+        case .dismissing:
+            current = nil
+            timer = nil
+            noticePresenter.cancel()
+            completion?()
+        }
```

Comment:
Let's chat over Slack regarding the general approach!!  I believe the actual problem could be addressed with really small changes.

Regardless of whatever we do, API(s) should **always** do exactly what their signature imply. In this particular case:

1. Dismiss might `presentNext`
2. (OR) actually dismiss
3. (OR) invoke completion

That means that the Dismiss method is tightly coupled with the `state`, and has unpredictable behavior.
Implementing this in an FSM approach could look like this:

```
private var state: State = .inactive {
   didSet {
      stateWasUpdated(newState: state)
   }
}
```

```
func stateWasUpdated(state: State) {
    switch state {
        case .presenting(let notice: Notice):
            // present the actual notice
        case .dismissing:
           // dismiss
    }
}
```

----

An FSM involves **States** and **Transitions** between them. Along with these lines, you would usually see an API that determines what the next state should be.